### ISSUE TITLE
pistackmond crashes immediately after previous instance terminated with sigkill

### ISSUE THREAD

**Issue by tomek-szczesny** [tomek-szczesny/pistackmon#6]
After pistackmond has been terminated non-gracefully (with sigkill), it appears the shared memory file still exists and stands in the way.
Sudoer can remove this file and later everything works as expected.

I'm not sure whether pistackmond should check if shared memory file exists and if so, maybe try deleting it. If that succeeds, then it means there is no other pistackmond process that keeps this file open. I'm just speculating here and would appreciate your input, @bablokb. Thanks!

```
mctom@odroidm1:~/git/pistackmon/sw$ sudo systemctl status pistackmond.service 
● pistackmond.service - PiStackMon Daemon
     Loaded: loaded (/etc/systemd/system/pistackmond.service; enabled; vendor preset: enable>
     Active: failed (Result: exit-code) since Sun 2022-07-03 13:52:14 CEST; 31s ago
    Process: 3353 ExecStart=/usr/local/sbin/pistackmond (code=exited, status=3)
   Main PID: 3353 (code=exited, status=3)

lip 03 13:52:14 odroidm1 systemd[1]: Started PiStackMon Daemon.
lip 03 13:52:14 odroidm1 pistackmond[3353]: shm_open failed: File exists
lip 03 13:52:14 odroidm1 systemd[1]: pistackmond.service: Main process exited, code=exited, >
lip 03 13:52:14 odroidm1 systemd[1]: pistackmond.service: Failed with result 'exit-code'.
...skipping...
● pistackmond.service - PiStackMon Daemon
     Loaded: loaded (/etc/systemd/system/pistackmond.service; enabled; vendor preset: enable>
     Active: failed (Result: exit-code) since Sun 2022-07-03 13:52:14 CEST; 31s ago
    Process: 3353 ExecStart=/usr/local/sbin/pistackmond (code=exited, status=3)
   Main PID: 3353 (code=exited, status=3)

lip 03 13:52:14 odroidm1 systemd[1]: Started PiStackMon Daemon.
lip 03 13:52:14 odroidm1 pistackmond[3353]: shm_open failed: File exists
lip 03 13:52:14 odroidm1 systemd[1]: pistackmond.service: Main process exited, code=exited, >
lip 03 13:52:14 odroidm1 systemd[1]: pistackmond.service: Failed with result 'exit-code'.
```

**Comment by bablokb** [comment]
Oh yes. The service creates the shared memory and therefore is also responsible for the deletion. I will fix this. The signal-handler are already in place.

**Comment by bablokb** [comment]
Should be fixed with commit 491d5954197e8289864cbb2f41f7fcfb871be11b

**Comment by bablokb** [comment]
> I don't think we should handle more signals, actually SIGKILL was very handy to kill pistackmond when I made it to loop infinitely by mistake. All the other signals got caught and waited for PWM thread to finish, which unfortunately was looped forever.

We can't trap SIGKILL anyhow.

> Ideally, pistackmond should give up and exit if it finds evidence of another instance already running. Otherwise, create the shared memory file or reuse the existing one (with any value already stored in there). I hope you agree this is the best scenario. :)
> 
> In bash world, I'd check if this command returns more than "1": `ps aux | grep pistackmond | grep root | wc -l` But perhaps more elegant solution exists in C++, can't tell really.

I think there is something more elegant in bash too ;-) (e.g. using -u and -C options of ps). But the main point is: does it make sense to blow up the code to catch a border-case which typically will only happen on a developer-machine?

So basically we have a few options:

  - *No change to the code, i.e. do nothing*. In this case pistackmond just won't start a second time because it assumes a running service. Normally this should be correct. If root is SIGKILLing the process, root should also remove the shm.
  - *Reuse the existing shm*. This could lead to two running services, but only if at least one of the pistackmond-executables is started manually. If you start via systemctl this should always be safe.

The caveat of the second option is the case where a user wants to set the user-led and just forgets to add the necessary parameter. But since we want to add cmdline options anyhow, we could add the option `-s` as a prereq for starting the service.

So I would suggest to leave the code as is for now. After adding cmdline-options, then we could switch to reusing the shm if it already exists. I feel no pity if someone afterwards explicitly starts the service twice.

**Comment by tomek-szczesny** [comment]
Yeah, I guess you're right, it's not critical by any means. If it was fixable with 5 lines of code that would be worth consideration, but otherwise I guess we can assume the service protects against running two instances.

> The caveat of the second option is the case where a user wants to set the user-led and just forgets to add the necessary parameter.

If run as non-root user it will crash anyway. There is a TODO in the code to catch that and yell at the user instead ;)

**Comment by bablokb** [comment]
Good. So I will now merge the user_led branch into main, then start with a new branch for the cmdline-options. Once this is in place I will switch the behavior to reuse the shm.

**Comment by tomek-szczesny** [comment]
Excellent. 
I'm leaving for Masters of Rock festival tomorrow morning so I won't be too responsive for the following week. :)
